File: android/app/src/main/java/com/hronro/jdime/ui/KeyboardView.kt
// Port of ios/Keyboard/UI/KeyboardView.swift. The full keyboard surface:
// candidate bar + key plane, with layer switching, shift state, theming, the
// expandable candidate grid, and lazy pagination. Driven by an InputSession.
package com.hronro.jdime.ui

import android.annotation.SuppressLint
import android.content.Context
import android.os.SystemClock
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.hronro.jdime.engine.Candidate
import com.hronro.jdime.engine.KeyAction
import com.hronro.jdime.engine.QuerySnapshot

@SuppressLint("ViewConstructor")
class KeyboardView(
    context: Context,
    val session: com.hronro.jdime.engine.InputSession,
    private var theme: KeyboardTheme,
) : FrameLayout(context) {

    /** Return key: owner decides commit-raw vs. newline (it knows the host). */
    var onReturn: (() -> Unit)? = null

    /** Localized label for the return key (set from the host's IME action). */
    var returnLabel: String = "换行"
        set(value) { field = value; keyGrid.returnLabel = value }

    private val density = context.resources.displayMetrics.density
    private val idiom: KeyboardIdiom =
        if (context.resources.configuration.smallestScreenWidthDp >= 600) KeyboardIdiom.PAD else KeyboardIdiom.PHONE
    private val compactHeight: Boolean =
        context.resources.configuration.orientation == android.content.res.Configuration.ORIENTATION_LANDSCAPE

    private var layer: KeyboardLayer = KeyboardLayer.LETTERS
    private var shift: ShiftState = ShiftState.OFF
    private var lastShiftTap: Long = 0

    private val container = LinearLayout(context).apply { orientation = LinearLayout.VERTICAL }
    private val candidateBar = CandidateBarView(context, theme)
    private val keyGrid = KeyboardLayoutView(context, theme, idiom)
    private var gridOverlay: CandidateGridView? = null

    /** Accumulated candidates for the current composition (across loaded pages). */
    private var items: List<Candidate> = emptyList()

    /** Bottom system-bar (gesture/nav) inset; the keyboard grows by this so keys clear it. */
    private var bottomInset = 0

    val preferredHeightPx: Int
        get() = ((CandidateBarView.HEIGHT_DP + KeyLayout.keysHeightDp(idiom, compactHeight)) * density).toInt()

    init {
        setBackgroundColor(theme.keyboardBackground)
        clipChildren = false

        keyGrid.onKey = { cap -> handle(cap) }
        candidateBar.onSelect = { i -> select(i) }
        candidateBar.onExpand = { expandGrid() }
        candidateBar.onNeedMore = { loadMore() }
        session.onChange = { snap, raw -> renderCandidates(snap, raw) }

        container.addView(
            candidateBar,
            LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, (CandidateBarView.HEIGHT_DP * density).toInt()),
        )
        container.addView(
            keyGrid,
            LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f),
        )
        addView(container, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))

        // Targeting SDK 35, the IME window is edge-to-edge and draws behind the system
        // navigation/gesture bar. Consume that bottom inset and pad it out (the keyboard
        // grows; the strip behind the gesture bar shows the keyboard background).
        ViewCompat.setOnApplyWindowInsetsListener(this) { _, insets ->
            // navigationBars() under-reports in gesture nav (just the home pill); the
            // system's IME nav bar that hosts the hide/switcher buttons is captured by
            // tappableElement(). Pad by the larger so the bottom row always clears it.
            val nav = insets.getInsets(WindowInsetsCompat.Type.navigationBars()).bottom
            val tap = insets.getInsets(WindowInsetsCompat.Type.tappableElement()).bottom
            val b = maxOf(nav, tap)
            if (b != bottomInset) {
                bottomInset = b
                setPadding(0, 0, 0, b)
                requestLayout()
            }
            insets
        }

        rebuildKeys()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        ViewCompat.requestApplyInsets(this)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val w = MeasureSpec.getSize(widthMeasureSpec)
        val h = preferredHeightPx + bottomInset
        super.onMeasure(
            MeasureSpec.makeMeasureSpec(w, MeasureSpec.EXACTLY),
            MeasureSpec.makeMeasureSpec(h, MeasureSpec.EXACTLY),
        )
        setMeasuredDimension(w, h)
    }

    fun applyTheme(theme: KeyboardTheme) {
        this.theme = theme
        setBackgroundColor(theme.keyboardBackground)
        candidateBar.apply(theme)
        keyGrid.apply(theme)
        gridOverlay?.apply(theme)
    }

    // MARK: - Keys

    private fun rebuildKeys() {
        keyGrid.setRows(KeyLayout.rows(layer, idiom))
        keyGrid.updateShift(shift)
    }

    private fun handle(cap: KeyCap) {
        when (cap) {
            is KeyCap.Char -> sendChar(cap.byte)
            is KeyCap.InsertLiteral -> { collapseGrid(); session.insertLiteral(cap.text) }
            KeyCap.Backspace -> { collapseGrid(); session.handle(KeyAction.Backspace) }
            KeyCap.Space -> { collapseGrid(); session.handle(KeyAction.EngineKey(0x20)) }
            KeyCap.Return -> { collapseGrid(); onReturn?.invoke() }
            KeyCap.Globe -> {}
            KeyCap.Shift -> toggleShift()
            is KeyCap.ToLayer -> setLayer(cap.layer)
            KeyCap.Spacer -> {}
        }
    }

    private fun sendChar(b: Byte) {
        var byte = b.toInt() and 0xFF
        if (shift != ShiftState.OFF && byte in 0x61..0x7A) byte -= 0x20
        collapseGrid()
        session.handle(KeyAction.EngineKey(byte.toByte()))
        if (shift == ShiftState.ONE_SHOT) { shift = ShiftState.OFF; keyGrid.updateShift(shift) }
    }

    private fun toggleShift() {
        val now = SystemClock.uptimeMillis()
        val doubleTap = (now - lastShiftTap) < 300
        lastShiftTap = now
        shift = when (shift) {
            ShiftState.OFF -> ShiftState.ONE_SHOT
            ShiftState.ONE_SHOT -> if (doubleTap) ShiftState.LOCKED else ShiftState.OFF
            ShiftState.LOCKED -> ShiftState.OFF
        }
        keyGrid.updateShift(shift)
    }

    private fun setLayer(l: KeyboardLayer) {
        layer = l
        shift = ShiftState.OFF
        rebuildKeys()
    }

    // MARK: - Candidates

    private fun renderCandidates(snap: QuerySnapshot, raw: String) {
        items = snap.options
        candidateBar.reset(raw, items, canExpand = snap.totalPages > 1)
        gridOverlay?.let { grid ->
            if (raw.isEmpty()) collapseGrid() else grid.setItems(items)
        }
    }

    private fun loadMore() {
        val more = session.loadMoreCandidates() ?: return
        if (more.isEmpty()) return
        items = items + more
        candidateBar.append(more)
        gridOverlay?.append(more)
    }

    private fun select(index: Int) {
        if (index < 0 || index >= items.size) return
        collapseGrid()
        session.commitCandidate(items[index].value)
    }

    private fun expandGrid() {
        if (gridOverlay != null || items.isEmpty()) return
        val grid = CandidateGridView(context, theme)
        grid.setItems(items)
        grid.onSelect = { i -> select(i) }
        grid.onNeedMore = { loadMore() }
        grid.onClose = { collapseGrid() }
        addView(grid, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        gridOverlay = grid
    }

    private fun collapseGrid() {
        gridOverlay?.let { removeView(it) }
        gridOverlay = null
    }
}

File: ios/Keyboard/UI/KeyboardView.swift
import UIKit

/// The full keyboard surface: candidate bar + key plane, with layer switching,
/// shift state, theming, the expandable candidate grid, and lazy pagination.
/// Driven by an `InputSession`; used by both the extension and the in-app preview.
final class KeyboardView: UIView {
    let session: InputSession

    /// Return key: owner decides commit-raw vs. newline (it knows the host).
    var onReturn: (() -> Void)?
    /// Globe key: owner advances to the next keyboard (extension only).
    var onNextKeyboard: (() -> Void)?
    /// Called when the preferred height changes (rotation / idiom change).
    var onHeightChanged: (() -> Void)?

    var showsNextKeyboardKey = true { didSet { if showsNextKeyboardKey != oldValue { rebuildKeys() } } }

    /// Localized label for the return key (set from the host's returnKeyType).
    /// Guarded: the setter walks every key, and the owner re-asserts it on
    /// every layout pass (see KeyboardViewController.viewWillLayoutSubviews).
    var returnLabel: String = "换行" {
        didSet { if returnLabel != oldValue { keyGrid.returnLabel = returnLabel } }
    }

    private(set) var theme: KeyboardTheme
    private var idiom: KeyboardIdiom = .phone { didSet { if idiom != oldValue { rebuildKeys() } } }
    private var compactHeight = false

    private var layer_: KeyboardLayer = .letters
    private var shift: ShiftState = .off
    private var lastShiftTap: CFTimeInterval = 0

    private let candidateBar: CandidateBarView
    private let keyGrid: KeyboardLayoutView
    private var gridOverlay: CandidateGridView?

    /// Accumulated candidates for the current composition (across loaded pages).
    private var items: [Candidate] = []

    var preferredHeight: CGFloat {
        CandidateBarView.height + KeyLayout.keysHeight(idiom: idiom, compactHeight: compactHeight)
    }

    init(session: InputSession, theme: KeyboardTheme = .light) {
        self.session = session
        self.theme = theme
        self.candidateBar = CandidateBarView(theme: theme)
        self.keyGrid = KeyboardLayoutView(theme: theme, idiom: .phone)
        super.init(frame: .zero)

        clipsToBounds = false
        backgroundColor = theme.keyboardBackground
        syncAppearance()

        keyGrid.popupHost = self
        keyGrid.clipsToBounds = false
        keyGrid.onKey = { [weak self] cap in self?.handle(cap) }

        candidateBar.onSelect = { [weak self] i in self?.select(i) }
        candidateBar.onExpand = { [weak self] in self?.expandGrid() }
        candidateBar.onNeedMore = { [weak self] in self?.loadMore() }

        session.onChange = { [weak self] snap, raw in self?.renderCandidates(snap, raw) }

        candidateBar.translatesAutoresizingMaskIntoConstraints = false
        keyGrid.translatesAutoresizingMaskIntoConstraints = false
        addSubview(candidateBar)
        addSubview(keyGrid)
        // Keep content inside the safe area so the notch / rounded corners don't
        // cover the left column or first candidate in landscape (no-op in portrait,
        // where the horizontal insets are 0). The background still fills edge-to-edge.
        let safe = safeAreaLayoutGuide
        NSLayoutConstraint.activate([
            candidateBar.topAnchor.constraint(equalTo: topAnchor),
            candidateBar.leadingAnchor.constraint(equalTo: safe.leadingAnchor),
            candidateBar.trailingAnchor.constraint(equalTo: safe.trailingAnchor),
            candidateBar.heightAnchor.constraint(equalToConstant: CandidateBarView.height),
            keyGrid.topAnchor.constraint(equalTo: candidateBar.bottomAnchor),
            keyGrid.leadingAnchor.constraint(equalTo: safe.leadingAnchor),
            keyGrid.trailingAnchor.constraint(equalTo: safe.trailingAnchor),
            keyGrid.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])

        rebuildKeys()
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    // MARK: - Traits

    override func didMoveToWindow() {
        super.didMoveToWindow()
        updateForTraits()
    }

    override func traitCollectionDidChange(_ previous: UITraitCollection?) {
        super.traitCollectionDidChange(previous)
        updateForTraits()
    }

    private func updateForTraits() {
        let newIdiom: KeyboardIdiom = traitCollection.userInterfaceIdiom == .pad ? .pad : .phone
        let newCompact = traitCollection.verticalSizeClass == .compact
        let changed = newIdiom != idiom || newCompact != compactHeight
        idiom = newIdiom
        compactHeight = newCompact
        if changed { onHeightChanged?() }
    }

    /// Switch the visible plane (QA / preview convenience).
    func showLayer(_ layer: KeyboardLayer) { setLayer(layer) }

    /// Expand the candidate grid (QA / preview convenience, mirrors the chevron).
    func expandCandidates() { expandGrid() }

    /// Show a character key's pressed state + preview bubble (QA / preview
    /// convenience — popups only live during a touch, so screenshots need this).
    func showKeyPopup(_ ch: Character) {
        guard let byte = ch.asciiValue else { return }
        keyGrid.subviews.compactMap { $0 as? KeyButton }
            .first { $0.spec.cap == .char(byte) }?
            .showPressedForQA()
    }

    func applyTheme(_ theme: KeyboardTheme) {
        // The owner re-applies on every layout pass (the extension's trait
        // updates can lag a live appearance flip, so layout is the only
        // reliable hook) — the guard makes that ~free.
        guard theme != self.theme else { return }
        self.theme = theme
        backgroundColor = theme.keyboardBackground
        syncAppearance()
        candidateBar.apply(theme: theme)
        keyGrid.apply(theme: theme)
        gridOverlay?.apply(theme: theme)
    }

    private func syncAppearance() {
        // Mirror the resolved appearance into this subtree's traits, so
        // trait-driven chrome (the glass popup) follows the HOST's requested
        // keyboardAppearance even when it differs from the system style.
        // Scoped to the keyboard view — overriding the controller's view
        // would feed the override back into the next `resolve(traits:)` and
        // wedge the theme after the host reverts to `.default`. No-op
        // visually for classic (its colors are explicit).
        //
        // NB: liquid glass draws NO background of its own
        // (`keyboardBackground` is `.clear`): the extension rides the system
        // keyboard panel's material, which also covers the strip above and
        // the globe/mic chin below our view — any material of ours would tint
        // just our rectangle and show as a seam against them. The in-app
        // preview, which has no system panel, supplies its own stand-in
        // (see KeyboardPreviewViewController).
        overrideUserInterfaceStyle = theme.isDark ? .dark : .light
    }

    // MARK: - Keys

    private func rebuildKeys() {
        keyGrid.setRows(KeyLayout.rows(layer: layer_, idiom: idiom, showGlobe: showsNextKeyboardKey))
        keyGrid.updateShift(shift)
    }

    private func handle(_ cap: KeyCap) {
        switch cap {
        case .char(let b):           sendChar(b)
        case .insertLiteral(let s):  collapseGrid(); session.insertLiteral(s)
        case .backspace:         collapseGrid(); session.handle(.backspace)
        case .space:             collapseGrid(); session.handle(.engineKey(0x20))
        case .ret:               collapseGrid(); onReturn?()
        case .globe:             onNextKeyboard?()
        case .shift:             toggleShift()
        case .toLayer(let l):    setLayer(l)
        case .spacer:            break   // no button is created for spacers
        }
    }

    private func sendChar(_ b: UInt8) {
        var byte = b
        if shift != .off, (0x61...0x7A).contains(b) { byte = b - 0x20 }
        collapseGrid()
        session.handle(.engineKey(byte))
        if shift == .oneShot { shift = .off; keyGrid.updateShift(shift) }
    }

    private func toggleShift() {
        let now = CACurrentMediaTime()
        let doubleTap = (now - lastShiftTap) < 0.3
        lastShiftTap = now
        switch shift {
        case .off:     shift = .oneShot
        case .oneShot: shift = doubleTap ? .locked : .off
        case .locked:  shift = .off
        }
        keyGrid.updateShift(shift)
    }

    private func setLayer(_ l: KeyboardLayer) {
        layer_ = l
        shift = .off
        rebuildKeys()
    }

    // MARK: - Candidates

    private func renderCandidates(_ snap: QuerySnapshot, _ raw: String) {
        items = snap.options
        // Chevron keyed to a fixed count, not totalPages, so the expand affordance
        // doesn't vanish for mid-size candidate sets when the engine page size grows.
        candidateBar.reset(composing: raw, items: items, canExpand: snap.optionsCount > 9)
        if let grid = gridOverlay {
            if raw.isEmpty { collapseGrid() } else { grid.setItems(items) }
        }
    }

    private func loadMore() {
        guard let more = session.loadMoreCandidates(), !more.isEmpty else { return }
        items.append(contentsOf: more)
        candidateBar.append(more)
        gridOverlay?.append(more)
    }

    private func select(_ index: Int) {
        guard index >= 0, index < items.count else { return }
        collapseGrid()
        session.commitCandidate(items[index].value)
    }

    private func expandGrid() {
        guard gridOverlay == nil, !items.isEmpty else { return }
        let grid = CandidateGridView(theme: theme)
        grid.setItems(items)
        grid.onSelect = { [weak self] i in self?.select(i) }
        grid.onNeedMore = { [weak self] in self?.loadMore() }
        grid.onClose = { [weak self] in self?.collapseGrid() }
        grid.translatesAutoresizingMaskIntoConstraints = false
        addSubview(grid)
        NSLayoutConstraint.activate([
            grid.topAnchor.constraint(equalTo: candidateBar.topAnchor),
            grid.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor),
            grid.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor),
            grid.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])
        gridOverlay = grid
        // The overlay is translucent in liquid glass (the material shows through),
        // so the plane and bar must actually vanish, not merely be covered.
        keyGrid.isHidden = true
        candidateBar.isHidden = true
    }

    private func collapseGrid() {
        gridOverlay?.removeFromSuperview()
        gridOverlay = nil
        keyGrid.isHidden = false
        candidateBar.isHidden = false
    }
}
